### ISSUE TITLE
FF "breakCircularDependency": is removed by amplify, re-adding results  into error on push 

### ISSUE THREAD

**Issue by parvusville** [aws-amplify/amplify-cli#12690]
### How did you install the Amplify CLI?

npm

### Amplify CLI Version

12.0.0

### What operating system are you using?

Pop Os

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

no

### Describe the bug

Yesterday I re-ran auth flow to add `preTokenGenerator` lambda, in order to Override ID Token Claims. I also enabled the `breakCircularDependency` feature flag in my `cli.json` to be able to add AppSync permissions to the `preTokenGenerator` lambda later. 

Now today when I tried to test stuff with amplify push, the push fails. CI/CD  builds are working fine for now. Doing amplify pull gets me bunch of stuff related to the `preTokenGenerator` lambda, that was not yet in my code base and I originally just commited and pushed the changes  to the CICD. Among these changes I see the `breakCircularDependency` FF removed in cli`.json`. It is still present in my `cli-inputs.json` though. 

If I now re-add the FF, I get this  `Reason: 1 validation error detected: Value 'functionmyAppa1b2abcd12PreTokenGeneration' at 'lambdaConfig.preTokenGeneration' failed to satisfy constraint: Member must satisfy regular expression pattern: arn:[\w+=/,.@-]+:[\w+=/,.@-]+:([\w+=/,.@-]*)?:[0-9]+:[\w+=/,.@-]+(:[\w+=/,.@-]+)?(:[\w+=/,.@-]+)?` error on push. (more details on reproduction section). 

`amplify push` seems to work as long as I dont re-add the breakCircularDependency, but now I am not able to add AppSync permissions! 

**I would gladly jump onto a discord call** to get this solved quickly. This feels like one of those things that takes way too long, when discussed here on the ticket on 1 msg per day basis. **This is a major blocker to our development**

### Expected behavior

breakCircularDependency FF should not dissapear from cli.json 

### Reproduction steps

This happens in dev  environment

1. add `"breakCircularDependency": true`  to `cli.json` 
2. `amplify update auth` run the custom flow keeping the old values, but enabling ID claim overrides and therefore adding `preTokenGenerator`  function
3. commit changes to Git, Amplify CI/CD runs and applies changes, everything works smoothly
4. I noticed that when switching from other testing environment back to dev (`amplify env checkout dev`) , stuff was added to `amplify-dependent-resources-ref.d.ts`  . I commit these into repository and move on
```
+    "myAppa1b2abcd12PreTokenGeneration": {
+      "Arn": "string",
+      "LambdaExecutionRole": "string",
+      "LambdaExecutionRoleArn": "string",
+      "Name": "string",
+      "Region": "string"
+    },
```

5. Now when I was trying to test other changes by doing a manual amplify push , I get following error:
```
🛑 The following resources failed to deploy:
Resource Name: UserPool (AWS::Cognito::UserPool)
Event Type: update
Reason: 1 validation error detected: Value 'functionmyAppa1b2abcd12PreTokenGeneration' at 'lambdaConfig.preTokenGeneration' failed to satisfy constraint: Member must satisfy regular expression pattern: arn:[\w+=/,.@-]+:[\w+=/,.@-]+:([\w+=/,.@-]*)?:[0-9]+:[\w+=/,.@-]+(:[\w+=/,.@-]+)?(:[\w+=/,.@-]+)? (Service: AWSCognitoIdentityProviderService; Status Code: 400; Error Code: InvalidParameterException; Request ID: 1cc49481-3716-4e5e-b657-802d8cbb20ef; Proxy: null)
URL: https://console.aws.amazon.com/cloudformation/home?region=eu-west-1#/stacks/arn%3Aaws%3Acloudformation%3Aeu-west-1%3A1234567%3Astack%2Famplify-myApp-dev-160123-authmyAppa1b2abcd12-1BMTCKBXON8LY%2F408ffe20-94b6-11ec-bad9-029788b6dda7/events


🛑 Resource is not in the state stackUpdateComplete
Name: UserPool (AWS::Cognito::UserPool), Event Type: update, Reason: 1 validation error detected: Value 'functionmyAppa1b2abcd12PreTokenGenerationArn' at 'lambdaConfig.preTokenGeneration' failed to satisfy constraint: Member must satisfy regular expression pattern: arn:[\w+=/,.@-]+:[\w+=/,.@-]+:([\w+=/,.@-]*)?:[0-9]+:[\w+=/,.@-]+(:[\w+=/,.@-]+)?(:[\w+=/,.@-]+)? (Service: AWSCognitoIdentityProviderService; Status Code: 400; Error Code: InvalidParameterException; Request ID: 1cc49481-3716-4e5e-b657-802d8cbb20ef; Proxy: null), IsCustomResource: false
```

6. Doing `amplify status`  shows me that there are changes to the `preTokenGenerator` lambda . 
7. At this points builds at CI/CD are still working fine
8. `amplify pull`  and adding a space to other existing lambda to test. amplify push  goes fine. 
9. I try to add an extra row change to my `preTokenGenerator`  lambda `alter-clamis.js`  file to trigger change to the function. `amplify push`  goes through just fine. 
10. I notice that after `amplify pull` the `"breakCircularDependency": true`  has been removed from my `cli.json` . 
I have set this FF to be able to later add AppSync permissions for my `preTokenGeneratorLambda` 
If I at this point try to add the AppSync permissions to the lambda via CLI, I get the Circular Dependency error.

11. If I re-add  `"breakCircularDependency": true`  to my `cli.json`  and try to push, I get the same error **from step 5**. 
12. I try to run `amplify update auth`  again (https://github.com/aws-amplify/amplify-cli/issues/4568#issuecomment-859690764), but `amplify push` still fails with the same error.

### Project Identifier

f59e77335650fce92677ec2970143a9b

### Log output

<details>

```
# Put your logs below this line


```

</details>

### Before submitting, please confirm:

- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

**Comment by parvusville** [comment]
**This seems to have solved the issue:**
After doing amplify pull, with breakCircularDependency removed from cli.sjon, I re-did amplify update auth and did amplify push. 
After this I re-added the breakCircularDependency, did amplify update auth again and now push goes through with the feature flag included. Also now I'm able to add the AppSync permissions.

I'm still curious of what caused this, and could this happen again? Maybe the fact that I did not do a local amplify push, and therefor I didn't have all amplify-generated resources in my repo? Then doing commits and amplify builds might have messed up something..

**Comment by ykethan** [comment]
Hey @parvusville,Thank you for reaching out. I was able to reproduce the issue with the following steps.

1. amplify init
2. amplify add auth and trigger with permissions to storage resource
3.push
4. amplify add env 
5. remove the flag in cli.json entirely
6. amplify push
7. then checkout to the previous env and run pull
8. observed the cli.json did not have the flag 
9. added the flag back in 
10. on the push observed the error.

Does this steps align to the existing reproduction. Was the flag disabled by removing the flag instead of setting the flag to false then was the environment checked out?

**Comment by parvusville** [comment]
@ykethan The initial description of my reproduction is the best memory I have of this event. But this does sound somewhat what I experienced, expect I never manually removed the flag from cli.json. As said in the previous message I got it working by removing and re-adding it, followed with `amplify update auth` and `amplify push`. Perhaps it might be a good idea to update these required steps in the documentation too?
